syntax = "proto3";

service CentralServer {
    rpc Heartbeat(Empty) returns (Empty) {}

    rpc RegisterClient(RegisterRequest) returns (RegisterResponse) {}
    rpc GetClientInfo(GetInfoRequest) returns (GetInfoResponse) {}
}

message Empty {}

message RegisterRequest {
    string username = 1;
    string ip = 2;
    int32 port = 3;
}

message RegisterResponse {
    bool success = 1;
    string body = 2;
}

message GetInfoRequest {
    string username = 1;
}

message GetInfoResponse {
    string ip = 1;
    int32 port = 2;
}

service ClientService {
    rpc Heartbeat(Empty) returns (Empty) {}

    rpc Connection(ConnectionRequest) returns (ConnectionResponse) {}

    rpc SendMessage(SendRequest) returns (Empty) {}
    rpc ReceiveMessage(ReceiveRequest) returns (Message) {}
}

message ConnectionRequest {
    string username = 1;
}

message ConnectionResponse {
    bool accept = 1;
}

message SendRequest {
    string username = 1;
    string message = 2;
}

message ReceiveRequest {
    string username = 1;
}

message Message {
    string body = 1;
}